### ISSUE TITLE
When Android battery saver mode is enabled, incoming MWA connections from Chrome fail

### ISSUE THREAD

**Issue by sdlaver** [solana-mobile/mobile-wallet-adapter#335]
When Battery Saver mode is enabled, Chrome web pages are paused immediately upon backgrounding. This breaks MWA, as the dapp is not able to communicate with the wallet.

Wallets should detect this case, and inform the user on why the connection failed, and how to address it (disable battery saver mode).

**Comment by steveluscher** [comment]
Oh dear.

So maybe we need to have that conversation, again, about [using the ability to autoplay a video or not](https://stackoverflow.com/a/71639385/802047) as a way to feature detect low power mode being on or off.

**Comment by ankur2136** [comment]
On Android its much simpler 
https://stackoverflow.com/a/31896873

**Comment by steveluscher** [comment]
> On Android its much simpler

I'm not sure this is a problem for native apps. Native apps aren't as aggressively paused as _webpages_ are.

There are no web APIs for detecting low power mode, because of fingerprinting & security concerns.

**Comment by sdlaver** [comment]
Proposed solution:

- walletlib detects this case using [isPowerSaveMode](https://developer.android.com/reference/android/os/PowerManager#isPowerSaveMode())
- If, after some number of seconds (default: 3), the wallet has not received an incoming connection, it will make a callback to the wallet indicating that the incoming connection may be suppressed by power saving mode.
- On receipt of the callback, the wallet can display appropriate UI to the user, and then return control to the dApp.

**Comment by steveluscher** [comment]
I think that's good iff this also affects native apps. If it's only web apps that are affected, we should probably find a way to prevent `wallet-adapter` from even making the call if it believes the device to be in low power mode.

> On receipt of the callback, the wallet can display appropriate UI to the user, and then return control to the dApp.

In this event, should we just invent and throw a protocol error and let the dApp handle it? Again, hesitant to do that if it's _only_ web apps that are affected.

**Comment by sdlaver** [comment]
Yes, it can affect native dApps as well. There's a fix for native dApps (on top of https://github.com/solana-mobile/mobile-wallet-adapter/issues/97) that involves giving a service handle for the backgrounded dApp to the wallet app.

> should we just invent and throw a protocol error and let the dApp handle it?
To handle it, the dapp needs to be able to receive the protocol error. For a web dapp, if the websocket connection is never established, there's no way for the wallet to give it any error or result codes.

**Comment by steveluscher** [comment]
Sorry, I'm not being clear.

1. By ‘affects web/native apps’ I mean ‘there is no mitigation.’ Because native apps have an API (?) to request some unpaused background time, I consider them unaffected by this problem.
2. > To handle it, the dapp needs to be able to receive the protocol error. For a web dapp, if the websocket connection is never established, there's no way for the wallet to give it any error or result codes.
  
    I meant to throw the protocol error _from the client_. Call `transact()` on the client. Client throws a `SolanaMobileWalletAdapterProtocolError` if the phone is found to be in low-power mode (ie. it doesn't even bother trying to associate).

**Comment by steveluscher** [comment]
Details on a proposed client-only solution, here: #343.

**Comment by Funkatronics** [comment]
Closing this issue because we have implemented a fix/warning in the wallet side SDK that most wallets have adopted by now. We dont have a client side solution implemented, but we can create a new issue or reopen this if we decide to pursue that.
